tsl/src/remote/connection.c

```diff
@@ -1995,145 +2246,53 @@ remote_result_close(PGresult *res)
 	PQclear(res);
 }
 
-/*
- * Cleanup connections and results at the end of a (sub-)transaction.
- *
- * This function is called at the end of transactions and sub-transactions to
- * auto-cleanup connections and result objects.
- */
-static void
-remote_connections_cleanup(SubTransactionId subtxid, bool isabort)
+bool
+remote_connection_set_single_row_mode(TSConnection *conn)
 {
-	ListNode *curr = connections.next;
-	unsigned int num_connections = 0;
-	unsigned int num_results = 0;
+	return PQsetSingleRowMode(conn->pg_conn) == 1;
+}
 
-	while (curr != &connections)
-	{
-		TSConnection *conn = (TSConnection *) curr;
+bool
+remote_connection_put_copy_data(const TSConnection *conn, const char *buffer, size_t nbytes,
+								TSConnectionError *err)
+{
+	int res = 0;
 
-		/* Move to next connection since closing the current one might
-		 * otherwise make the curr pointer invalid. */
-		curr = curr->next;
+	Assert(PQisnonblocking(conn->pg_conn));
+	Assert(conn->status == CONN_COPY_IN);
 
-		if (conn->autoclose && (subtxid == InvalidSubTransactionId || subtxid == conn->subtxid))
-		{
-			/* Closes the connection and frees all its PGresult objects */
-			remote_connection_close(conn);
-			num_connections++;
-		}
-		else
-		{
-			/* We're not closing the connection, but we should clean up any
-			 * lingering results */
-			ListNode *curr_result = conn->results.next;
+	do
+	{
+		res = PQputCopyData(conn->pg_conn, buffer, nbytes);
 
-			while (curr_result != &conn->results)
+		switch (res)
+		{
+			case 0:
 			{
-				ResultEntry *entry = (ResultEntry *) curr_result;
-
-				curr_result = curr_result->next;
-
-				if (subtxid == InvalidSubTransactionId || subtxid == entry->subtxid)
-				{
-					PQclear(entry->result);
-					num_results++;
-				}
+				/* Full buffers, wait... */
+				if (!wait_and_consume_input(conn, -1, WL_SOCKET_WRITEABLE))
```

Comment:
You don't need to call `PQflush()` after socket is ready for write?

Reply:
Yes, it happens at the end of this function. I guess it is not easy to see because of the weird diff.